virtcontainers/fc.go

```diff
@@ -0,0 +1,580 @@
+		return err
+	}
+
+	return nil
+}
+
+// Firecracker supports replacing the host drive used once the VM has booted up
+func (fc *firecracker) fcUpdateBlockDrive(drive config.BlockDrive) error {
+	span, _ := fc.trace("fcUpdateBlockDrive")
+	defer span.Finish()
+
+	driveID := drive.ID
+	driveParams := ops.NewPatchGuestDriveByIDParams()
+	driveParams.SetDriveID(driveID)
+
+	driveFc := &models.PartialDrive{
+		DriveID:    &driveID,
+		PathOnHost: &drive.File, //This is the only property that can be modified
+	}
+	driveParams.SetBody(driveFc)
+	_, err := fc.client.Operations.PatchGuestDriveByID(driveParams)
+	if err != nil {
+		fc.Logger().WithField("fcUpdateBlockDrive failed:", err).Debug()
+		return err
+	}
+
+	actionParams := ops.NewCreateSyncActionParams()
+	actionInfo := &models.InstanceActionInfo{
+		ActionType: "BlockDeviceRescan",
+	}
+	actionParams.SetInfo(actionInfo)
+	_, err = fc.client.Operations.CreateSyncAction(actionParams)
+	if err != nil {
+		fc.Logger().WithField("fcUpdateBlockDrive BlockDeviceRescan failed:", err).Debug()
+		return err
+	}
+
+	return nil
+}
+
+// addDevice will add extra devices to firecracker.  Limited to configure before the
+// virtual machine starts.  Devices include drivers and network interfaces only.
+func (fc *firecracker) addDevice(devInfo interface{}, devType deviceType) error {
+	span, _ := fc.trace("addDevice")
+	defer span.Finish()
+
+	fc.state.RLock()
+	defer fc.state.RUnlock()
+
+	if fc.state.state == notReady {
+		dev := firecrackerDevice{
+			dev:     devInfo,
+			devType: devType,
+		}
+		fc.Logger().Warn("FC not ready, queueing device")
+		fc.pendingDevices = append(fc.pendingDevices, dev)
+		return nil
+	}
+
+	fc.Logger().Info("adding a device")
+	switch v := devInfo.(type) {
+	case Endpoint:
+		fc.Logger().Info("Adding net device")
+		return fc.fcAddNetDevice(v)
+	case config.BlockDrive:
+		fc.Logger().Info("Adding block device")
+		return fc.fcAddBlockDrive(v)
+	case kataVSOCK:
+		fc.Logger().Info("Adding vsock")
+		return fc.fcAddVsock(v)
+	default:
+		fc.Logger().Warning("unknown device")
+		break
+	}
+
+	return nil
+}
+
+// hotplugAddDevice not supported in Firecracker VMM
+func (fc *firecracker) hotplugAddDevice(devInfo interface{}, devType deviceType) (interface{}, error) {
+	span, _ := fc.trace("hotplugAddDevice")
+	defer span.Finish()
+
+	switch devType {
+	case blockDev:
+		//The drive placeholder has to exist prior to Update
+		return nil, fc.fcUpdateBlockDrive(*devInfo.(*config.BlockDrive))
+	default:
+		fc.Logger().WithField("devInfo", devInfo).Warn("hotplugAddDevce: unknown device")
+		fc.Logger().WithField("devType:", devType).Warn("hotplugAddDevce: unknown device")
+		break
+	}
+
+	return nil, nil
+}
+
+// hotplugRemoveDevice not supported in Firecracker VMM
+func (fc *firecracker) hotplugRemoveDevice(devInfo interface{}, devType deviceType) (interface{}, error) {
+	return nil, fmt.Errorf("firecracker does not support hotplugRemoveDevice")
```

Comment:
I remember seeing `container.stop()` failure due to this when running fc. Maybe we should just print a warning here and return success?

Reply:
You are right. As firecracker does not support this feature. Will return from here without error.